### DISCUSSION TITLE
Improving bitrate & framerate

### DISCUSSION THREAD

**Discussion by master131** [tesla-android/issue-tracker#281] [Q&A] [answered]
Hey guys, I just used Tesla Android today for the first time and it works well (unfortunately due to being in Australia, the browser cannot be used except while in park - but not relevant to the question).

I noticed that the frame rate is not as smooth as possible (slightly choppy, guessing somewhere between 10-20fps) and there is visible color banding or blocking artifacts around certain text, indicative of perhaps low bitrate? I'm just wondering if this is some sort of hardware limitation or if there is still further optimisations/tweaking that can be done?

I'm running a Pi 4B 4GB inside a Argon One V2 case so thermal throttling shouldn't be a factor.

**Comment by mikegapinski** [comment, marked as answer]
This has to do with the video in drive bypass, it affects performance since I cannot use the video tag. With the responsive display being finished I can get something better with h264 for those using Tesla Android in park/fullscreen. 

Like always, there are always improvements that I can do, I am on this train for over 18 months now 😄

**Reply by master131** [reply]
I appreciate the fast response. That would be awesome if possible - it means that being able to use third party multimedia apps becomes much more better while in park (ie. Kodi, VLC, other streaming apps).

**Reply by dvrtech-us** [reply]
I was playing with ideas for just video playback, have you considered MPEG1 instead? I was exploring https://jsmpeg.co/

**Comment by mikegapinski** [comment]
It won’t be hardware accelerated so it makes 0 sense. I have a h264 based solution baked into the OS now but it was worse in terms of performance than mjpeg. Tesla has enabled acceleration for this in chrome and that is the only reason why it runs well

Wysyłane z aplikacji Outlook dla systemu iOS<https://aka.ms/o0ukef>
________________________________
Od: DVRTech ***@***.***>
Wysłane: Saturday, August 26, 2023 10:46:48 PM
Do: tesla-android/issue-tracker ***@***.***>
DW: Michał Gapiński ***@***.***>; Comment ***@***.***>
Temat: Re: [tesla-android/issue-tracker] Improving bitrate & framerate (Discussion #281)


I was playing with ideas for just video playback, have you considered MPEG1 instead? I was exploring https://jsmpeg.co/

**Comment by mikegapinski** [comment]
The next performance breakthrough is on the pi side in the gpu drivers and their virtual display implementation. It won’t gain much more fps- we are reaching a limit of what you can do without the video tag and that is the only thing that works in motion. I tested every possible combination in the last two years 😅
Wysyłane z aplikacji Outlook dla systemu iOS<https://aka.ms/o0ukef>
________________________________
Od: Michał Gapiński ***@***.***>
Wysłane: Saturday, August 26, 2023 10:48:39 PM
Do: tesla-android/issue-tracker ***@***.***>; tesla-android/issue-tracker ***@***.***>
DW: Michał Gapiński ***@***.***>; Comment ***@***.***>
Temat: Re: [tesla-android/issue-tracker] Improving bitrate & framerate (Discussion #281)

It won’t be hardware accelerated so it makes 0 sense. I have a h264 based solution baked into the OS now but it was worse in terms of performance than mjpeg. Tesla has enabled acceleration for this in chrome and that is the only reason why it runs well

Wysyłane z aplikacji Outlook dla systemu iOS<https://aka.ms/o0ukef>
________________________________
Od: DVRTech ***@***.***>
Wysłane: Saturday, August 26, 2023 10:46:48 PM
Do: tesla-android/issue-tracker ***@***.***>
DW: Michał Gapiński ***@***.***>; Comment ***@***.***>
Temat: Re: [tesla-android/issue-tracker] Improving bitrate & framerate (Discussion #281)


I was playing with ideas for just video playback, have you considered MPEG1 instead? I was exploring https://jsmpeg.co/

**Reply by dvrtech-us** [reply]
That’s cool I didn’t realize you could achieve acceleration on the Tesla side without the video tag, I assumed you were using the JavaScript MPEG-4 decoder to do in motion decoding. I guess I should take more time to read the code. I definitely plan to fork and contribute as time allows. I was also pondering the idea of using an actual cell phone, so that there is data service on a standard plan. Have you ever played with that concept?

**Reply by mikegapinski** [reply]
If you want to do something like TA on a handheld device don't bother. It will melt down after 10 minutes of use.

You could in theory but you'd need a pixel with a custom build of android and some framework changes. By default the performance of the hotspot is poor
